### PR TITLE
Stabilize covariance learning with `FillScaleTriL` and update config behaviour

### PR DESCRIPTION
**Pull request by patel-zeel**
## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [x] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] Documentation content changes
- [ ] Other (please describe):

## What is the current v/s new behavior?

Issue Number: #127

As discussed in #127, the following is the summary of changes:

1. Replace `tfb.FillTriangular` with `tfb.FillScaleTriL` in `config.py`
1. Add the "Custom transformations" section in `examples/uncollapsed_vi.pct.py` to illustrate the use of custom transformation for lower triangular Cholesky parameters. 
1. Change the way config is created currently. Replace `get_defaults()` with `get_global_config()`. Add `reset_global_config()` and `get_default_config()` methods.
1. Add the "Conversion between `.ipynb` and `.py`" section in README.md

### "Custom transformations" section in `examples/uncollapsed_vi.pct.py`
I am showing how to use a custom `triangular_transform` in `gpjax` (the method also generalizes for any transform).

```py
gpx_config = get_global_config()
transformations = gpx_config.transformations
jitter = gpx_config.jitter

triangular_transform = dx.Chain(
    [tfb.FillScaleTriL(diag_bijector=tfb.Square(), diag_shift=jnp.array(jitter))]
)

transformations.update({"triangular_transform": triangular_transform})
```

I have added a point that the `Square` bijector may lead to a faster convergence but can be unstable compared to `Softplus` bijector.

| Softplus | Square |
| :-: | :-: |
| image  |  image  |

#### How is it done currently?
1. `get_defaults` method returns the global config. If the config is unavailable, it creates one.

#### What is the new behavior?
1. `get_global_config` returns the global config. If the config is unavailable, it creates one. If JAX precision changes, it makes appropriate changes to the current global config e.g. update `FillScaleTriL` 
2. `get_default_config` returns the default config and does not update any global config.
3. `reset_global_config` resets global config to default config.

### Conversion between `.ipynb` and `.py`
The following quick commands are introduced in README.md to convert between `.ipynb` and `.py`.
image

> ### Some changes are attempted in this PR without a detailed prior discussion. Please consider those as suggestions and suggest your thoughts on the required modifications.

## What is not done?

I have not ensured yet that `get_global_config` does not get invoked at `import gpjax as gpx` (O1 in #127). Execution of https://github.com/JaxGaussianProcesses/GPJax/blob/c50d34d44977e3ef2290576ee1e9b82a69656431/gpjax/__init__.py#L16 runs [natural_gradients.py](https://github.com/JaxGaussianProcesses/GPJax/blob/master/gpjax/natural_gradients.py) which invokes `get_global_config`. Any suggestions on how to solve this?

### CODE REVIEW

**Comment by Daniel-Dodd** [comment]
Thanks, @patel-zeel! This PR looks great. Will start my review soon. 

> ## What is not done?
> I have not ensured yet that `get_global_config` does not get invoked at `import gpjax as gpx` (O1 in #127). Execution of
> 
> https://github.com/JaxGaussianProcesses/GPJax/blob/c50d34d44977e3ef2290576ee1e9b82a69656431/gpjax/__init__.py#L16
> 
> runs [natural_gradients.py](https://github.com/JaxGaussianProcesses/GPJax/blob/master/gpjax/natural_gradients.py) which invokes `get_global_config`. Any suggestions on how to solve this?

Is `get_global_config` invoked in `natural_gradients.py` solely via `DEFAULT_JITTER`? We can then move the jitter inside the relevant functions/objects as done in `gps.py` and `variational_families.py`, to resolve this.

**Review by Daniel-Dodd — CHANGES_REQUESTED** [review]
Thanks @patel-zeel, this is an excellent PR. Just a few minor things, and to see if O1 in https://github.com/JaxGaussianProcesses/GPJax/issues/127 might be resolved via removing the default jitter and define jitter inside objects/functions akin to `gps.py`/`variational_families.py`.

File: examples/uncollapsed_vi.pct.py [hunk 1]
```diff
@@ -191,9 +198,61 @@
 ]
 plt.show()
 
+# %% [markdown]
+# ## Custom transformations
+#
+# To train a covariance matrix, `gpjax` uses `tfb.FillScaleTriL` transformation by default. `tfb.FillScaleTriL` fills a 1d vector into a lower triangular matrix and then applies `SoftPlus` transformation on the diagonal to satisfy the necessary conditions for a valid Cholesky matrix. Users can change this default transformation with another valid transformation of their choice. For example, `Square` transformation on the diagonal can also serve the purpose.
```
[the remark this answers is not in the document]
Comment (patel-zeel) [reply examples/uncollapsed_vi.pct.py]: Done.

File: examples/uncollapsed_vi.pct.py [hunk 2]
```diff
@@ -191,9 +198,61 @@
...
+ax.fill_between(xtest.flatten(), meanf - sigma, meanf + sigma, alpha=0.3)
+[
+    ax.axvline(x=z_i, color="black", alpha=0.3, linewidth=1)
+    for z_i in learned_params["variational_family"]["inducing_inputs"]
+]
+plt.show()
+
+# %% [markdown]
+# We can see that `Square` transformation is able to get relatively better fit compared to `Softplus` with the same number of iterations, but `Softplus` is recommended over `Square` for stability of optimization.
+
```
[the remark this answers is not in the document]
Comment (patel-zeel) [reply examples/uncollapsed_vi.pct.py]: Done.

File: gpjax/config.py [hunk 3]
```diff
@@ -21,35 +22,88 @@
...
+    # Update the x64 state
+    x64_state = jax.config.x64_enabled
+    __config.x64_state = x64_state
+
+    # Update the x64 sensitive bijectors
+    FillScaleTriL = dx.Chain(
+        [
+            tfb.FillScaleTriL(diag_shift=jnp.array(__config.jitter)),
+        ]
+    )  # TODO: Dan to chain methods.
```
[the remark this answers is not in the document]
Comment (patel-zeel) [reply gpjax/config.py]: Done.

File: README.md [hunk 4]
```diff
@@ -48,6 +48,20 @@ Feel free to join our [Slack Channel](https://join.slack.com/t/gpjax/shared_invi
...
+
+```bash
+jupytext --to notebook example.py
+```
+
+* To convert `example.ipynb` to `example.py`, run:
+
+```bash
+jupytext --to py:percent example.ipynb
+```
```
[the remark this answers is not in the document]
Comment (patel-zeel) [reply README.md:64]: Thank you!

File: gpjax/config.py [hunk 5]
```diff
@@ -21,35 +22,88 @@
...
+    # If the global config is available, check if the x64 state has changed
+    x64_state = jax.config.x64_enabled
+
+    # If the x64 state has not changed, return the existing global config
+    if x64_state is __config.x64_state:
+        return __config
+
+    # If the x64 state has changed, return the updated global config
+    update_x64_sensitive_settings()
+    return __config
```
[the remark this answers is not in the document]
Comment (patel-zeel) [reply gpjax/config.py:58]: Thanks!

**Comment by patel-zeel** [comment]
Thank you for the review, @Daniel-Dodd. I have resolved O1 in #127 by moving the jitter inside the function. I have also addressed the other comments.

**Comment by patel-zeel** [comment]
@Daniel-Dodd since we have addressed O1 in #127, we may need to ensure it does not get violated in the future. I think it is hard to detect such things via unit testing. However, I have added a relevant test (`test_config_on_library_import`) in `test_config` which calls `get_global_config_if_exists` function to check if the global config exists. This test seems to be passing only because of two reasons: i) `test_config_on_library_import` test is defined on top and thus it is executed first within `test_config.py` ii) All tests running before `test_config.py` do not invoke the creation of global config. What are your thoughts on this?

**Comment by Daniel-Dodd** [comment]
Hi @patel-zeel, thank you for addressing my comments.

> @Daniel-Dodd since we have addressed O1 in #127, we may need to ensure it does not get violated in the future.

In agreement with this. However, the latest test workflow failed on `test_config_on_library_import` (https://github.com/JaxGaussianProcesses/GPJax/actions/runs/3729043958/jobs/6330357412)? Happy to merge this PR with this test removed for now, and open a separate issue regarding checks for O1 violation?

**Comment by patel-zeel** [comment]
Yes, that looks like a way to go :)

**Comment by Daniel-Dodd** [comment]
@patel-zeel Nice! Running the test workflow now, will merge soon as the tests pass. :)

**Review by Daniel-Dodd — APPROVED** [review]

**Pull request merged**